### PR TITLE
Guardian Labs: sponsor badge

### PR DESCRIPTION
**Pull request by alinaboghiu**
## What does this change?

This change introduces a sponsor's badge into the existing badge field of the DCRCollectionType.

The sponsor badge will be provided when: 
* all the cards in a container have the same sponsor and 
* the container palette is "branded", which means a "branded" tag was added to it in the fronts config tool 

Since the DCRContainerPalette needs to now be aware of the Branded type, I have updated decideContainerOverrides.ts to allow for the fact that in the case of a Branded container (or Labs container) the only override is the background colour.

## Why?

Guardian Labs containers will display this sponsor badge more prominently when it is provided. 

## Screenshots

Note: the badge is not yet displayed in the correct location, and it also needs to be removed from the individual cards. But one step at a time.. 

| Before      | After      |
| ----------- | ---------- |
| Screenshot 2023-05-22 at 15 32 10 | Screenshot 2023-05-22 at 15 31 29 |

### CODE REVIEW

**Review by AshCorr — APPROVED** [review]

File: dotcom-rendering/src/model/enhanceCollections.ts [hunk 1]
```diff
@@ -14,6 +20,22 @@ const FORBIDDEN_CONTAINERS = [
 const isSupported = (collection: FECollectionType): boolean =>
 	!FORBIDDEN_CONTAINERS.includes(collection.displayName);
 
+function getBrandingFromCards(
+	allCards: FEFrontCard[],
+	editionId: 'UK' | 'US' | 'AU' | 'INT' | 'EUR',
```
[the remark this answers is not in the document]
Comment (alinaboghiu) [reply dotcom-rendering/src/model/enhanceCollections.ts:25]: oh good spot thank you!

File: dotcom-rendering/src/model/enhanceCollections.ts [hunk 2]
```diff
@@ -14,6 +20,22 @@ const FORBIDDEN_CONTAINERS = [
...
+): (Branding | undefined)[] {
+	return allCards.map(
+		(card) =>
+			card.properties.editionBrandings.find(
+				(editionBranding) => editionBranding.edition.id === editionId,
+			)?.branding,
+	);
+}
+
+function allCardsHaveSponsors(allBranding: (Branding | undefined)[]): boolean {
```
[the remark this answers is not in the document]
Comment (alinaboghiu) [reply dotcom-rendering/src/model/enhanceCollections.ts:35]: I will try this, but I think I tried and it just didn't like me doing this at all :( 
might ask to pair if you have a minute

File: dotcom-rendering/src/model/article-schema.json [hunk 3]
```diff
@@ -4595,16 +4595,16 @@
...
+                "href": {
                     "type": "string"
                 }
             },
             "required": [
-                "imageUrl",
-                "seriesTag"
+                "href",
+                "imageSrc"
             ]
```
Comment (OllysCoding) [inline dotcom-rendering/src/model/article-schema.json:4608]: What's the reason for this change?

Comment (alinaboghiu) [reply dotcom-rendering/src/model/article-schema.json:4608]: Series tag isn't always a series tag, for paid content it's a link to the sponsor's page. So href felt more appropriate, since that value is used as the href for the link. 

I guess imageURL could have stayed the same but I thought imageScr is more descriptive and since I was changing the other thing.. But I could be easily persuaded to change that back

File: dotcom-rendering/src/model/enhanceCollections.ts [hunk 4]
```diff
@@ -26,6 +48,12 @@ export const enhanceCollections = (
 		const containerPalette = decideContainerPalette(
 			collection.config.metadata?.map((meta) => meta.type),
 		);
+		const allBranding = getBrandingFromCards(
+			[...collection.curated, ...collection.curated],
+			editionId,
+		);
+		const allCardsHaveBranding = allCardsHaveSponsors(allBranding);
```
Comment (OllysCoding) [inline dotcom-rendering/src/model/enhanceCollections.ts:55]: If the branding isn't used unless _all_ the cards have branding, perhaps `getBrandingFromCards` can return just the brandings (e.g `Branding[]` rather than `(Branding | Undefined)[]`) - by filtering on undefined.

Then you could simplify `allCardsHaveBranding = allBranding.length === allCards.length` (you'd need to define `const allCards = [...collection.curated, ...collection.curated]`

File: dotcom-rendering/src/model/enhanceCollections.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/model/enhanceCollections.ts:35]: There’s [`isNonNullable` from `@guardian/lib`](https://github.com/guardian/csnx/blob/a64bbc6d1fced5ced51c9295dc53163706855247/libs/%40guardian/libs/src/isNonNullable/isNonNullable.ts) which may help solve this?

You can do `allbranding.filter(isNonNullable)` and you’d get a `Branding[]`.

File: dotcom-rendering/src/model/enhanceCollections.ts (hunk 4 above)
Comment (mxdvl) [reply dotcom-rendering/src/model/enhanceCollections.ts:55]: There’s [`isNonNullable` from `@guardian/lib`](https://github.com/guardian/csnx/blob/a64bbc6d1fced5ced51c9295dc53163706855247/libs/%40guardian/libs/src/isNonNullable/isNonNullable.ts) which may help solve this?

```ts
const cardsWithBranding: Branding[] = allCards
  .map(() => { /* do stuff */ })
  .filter(isNonNullable);
```

File: dotcom-rendering/src/model/enhanceCollections.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (alinaboghiu) [reply dotcom-rendering/src/model/enhanceCollections.ts]: Resolved with Max's suggestion thanks everyone

File: dotcom-rendering/src/model/article-schema.json (hunk 3 above)
Comment (OllysCoding) [reply dotcom-rendering/src/model/article-schema.json:4608]: Okay! Is this something that needs to be updated in Frontend so the data comes through with the correct keys - or were they already set as both anyway?

File: dotcom-rendering/src/model/enhanceCollections.ts [hunk 5]
```diff
@@ -14,6 +21,28 @@ const FORBIDDEN_CONTAINERS = [
...
+		)
+		.filter(isNonNullable);
+}
+
+function allCardsHaveSponsors(
+	allCards: FEFrontCard[],
+	allBranding: Branding[],
+): boolean {
+	return allCards.length === allBranding.length;
+}
```
Comment (OllysCoding) [inline dotcom-rendering/src/model/enhanceCollections.ts:44]: Does this need to be its own function?

**Review by OllysCoding — APPROVED** [review]
+1 :)

File: dotcom-rendering/src/model/article-schema.json (hunk 3 above)
Comment (alinaboghiu) [reply dotcom-rendering/src/model/article-schema.json:4608]: was always being set 👍

File: dotcom-rendering/src/model/enhanceCollections.ts (hunk 5 above)
Comment (alinaboghiu) [reply dotcom-rendering/src/model/enhanceCollections.ts:44]: not anymore :D
